File: Week 3/python/13.reverse.py
"""
WEEK 3 - PYTHON LOOPS & NUMBER THEORY
Topic: Reverse the Digits of an Integer
File: 13.reverse.py

PROBLEM:
Read N, output the integer obtained by reversing its digits. Trailing zeros
in the original become leading zeros in the reversed number and are
dropped (10400 -> 401).

CONCEPT:
Standard digit-extraction loop:
  rev = rev * 10 + last_digit
  n  //= 10
Repeat until n == 0.

KEY POINTS:
 - Handle negative numbers: reverse abs(n), preserve sign.
 - Or: int(str(abs(n))[::-1]) -- string-based one-liner.
 - For 32-bit overflow check (LeetCode-style), validate against 2^31-1.

SYNTAX:
 rev = 0
 while n:
     rev = rev * 10 + n % 10
     n //= 10

DRY RUN (n=1234):
 rev=0; n=1234 -> rev=4,  n=123
 rev=4;          -> rev=43, n=12
 rev=43;         -> rev=432,n=1
 rev=432;        -> rev=4321,n=0
 print 4321

COMPLEXITY: O(d) where d = number of digits.
"""

import sys


def reverse_int(n: int) -> int:
    sign = -1 if n < 0 else 1
    n = abs(n)
    rev = 0
    while n > 0:
        rev = rev * 10 + n % 10
        n //= 10
    return sign * rev


def reverse_via_string(n: int) -> int:
    """One-liner via string slicing -- often the most Pythonic."""
    sign = -1 if n < 0 else 1
    return sign * int(str(abs(n))[::-1])


def main() -> None:
    if sys.stdin.isatty():
        for n in (1234, 10400, 7, -42):
            print(f"reverse({n:>6}) = {reverse_int(n)}    str-slice: {reverse_via_string(n)}")
        return
    n = int(sys.stdin.read().split()[0])
    print(reverse_int(n))


if __name__ == "__main__":
    main()


# NOTES:
# - String slicing `s[::-1]` is the Pythonic way to reverse any sequence.
# - For Java/C++, the digit-extraction loop is the only sane option (no easy reverse).
# - Mind the sign and trailing zeros.

File: Week 3/cpp/13.reverse.cpp
/*
 * WEEK 3 - C++ LOOPS & NUMBER THEORY
 * Topic: Reverse the Digits of an Integer
 * File: 13.reverse.cpp
 *
 * PROBLEM:
 *  Read N, output reversed integer (drop trailing zeros).
 *
 * KEY POINTS:
 *  - Standard digit-extraction loop.
 *  - Cast n to long long if you might overflow (e.g. 2147483647 reversed).
 *
 * DRY RUN:
 *  N = 1234 -> 4321
 *  N = 10400 -> 401
 */

#include <iostream>
#include <cstdlib>

long long reverse_int(long long n) {
    long long sign = (n < 0) ? -1 : 1;
    n = std::llabs(n);
    long long rev = 0;
    while (n > 0) {
        rev = rev * 10 + n % 10;
        n /= 10;
    }
    return sign * rev;
}

int main() {
    long long n;
    if (!(std::cin >> n)) {
        for (long long v : {1234LL, 10400LL, 7LL, -42LL}) {
            std::cout << v << " -> " << reverse_int(v) << '\n';
        }
        return 0;
    }
    std::cout << reverse_int(n) << '\n';
    return 0;
}

/*
 * NOTES:
 *  - For 32-bit safety (LeetCode), check rev > INT_MAX/10 before each multiply.
 *  - C++ has no built-in string-reverse trick that maintains type; use std::to_string
 *    + std::reverse for the string-based variant.
 */
